<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Tutorial &mdash; iDiffIR 0.3.1 documentation</title>
    
    <link rel="stylesheet" href="_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="iDiffIR 0.3.1 documentation" href="index.html" />
    <link rel="next" title="Contact" href="contact.html" />
    <link rel="prev" title="Get iDiffIR" href="installation.html" /> 
  </head>
  <body>
      <div class="header">
        <a href="index.html">
          <img class="logo" src="_static/logo.png" alt="Logo"/>
        </a>
      </div>
      <div class="topnav">
      
        <p>
        «&#160;&#160;<a href="installation.html">Get iDiffIR</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="contact.html">Contact</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h1>
<p>This tutorial is meant as a complete walk-through for identifying
differential IR from RNA-seq data.  The process is as follows:</p>
<ol class="arabic simple">
<li>Align reads</li>
<li>Filter alignments</li>
<li>Get read depths</li>
<li>Analyze depths for differential IR</li>
</ol>
<div class="section" id="align-reads">
<h2>Align reads<a class="headerlink" href="#align-reads" title="Permalink to this headline">¶</a></h2>
<p><strong>iDiffIR</strong> accepts any sorted, indexed BAM file for single- or
paired-end reads.  It may be helpful to use an aligner that
provides an &#8220;NH&#8221; tag (such as <a class="reference external" href="https://code.google.com/p/rna-star/">STAR</a> or <a class="reference external" href="http://ccb.jhu.edu/software/tophat/index.shtml">TopHat</a>)
to identify unique alignments; however, there are other ways of
removing multi-reads (reads mapping to more than
one genomic location).</p>
</div>
<div class="section" id="filtering-alignments">
<h2>Filtering alignments<a class="headerlink" href="#filtering-alignments" title="Permalink to this headline">¶</a></h2>
<p>It is good practice to remove multi-reads and filter spliced
alignments for false-positive junctions  before beginning
transcriptome analysis of RNA-Seq.  Assuming your alignments
are in a file named aligned.bam, run the following to extract
only uniquely aligned reads:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$samtools view -h aligned.bam | grep -we &#39;NH:i:1&#39; -e &#39;^@*&#39; &gt; unique.sam
</pre></div>
</div>
<p>You may wish to filter out potential false-positive splice junctions
using SpliceGrapher<a class="reference internal" href="#sg" id="id1">[SG]</a> <strong class="program">sam_filter.py</strong>
from uniquely aligned reads.  While the presence
of false-positive splice junctions may not adversely affect
identifying differential IR analysis, it is a critical step for
detecting all other forms of AS (exon skipping, alternative
acceptor/donor events).</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>Usage: sam_filter.py SAM/BAM-file classifiers [options]

Removes false-positive spliced alignments from a SAM or BAM file.
The classifers may be given as a zip-file or as a list of classifier
configuration (.cfg) files.

Options:
  -h, --help  show this help message and exit
  -C CIGAR    File for storing unrecognized CIGAR strings [default: none]
  -f FASTA    Reference genome FASTA [default: none]
  -F FPSITES  File for storing false-positive SAM alignment records [default:
              none]
  -m MODEL    Gene model file (GTF/GFF3 format) [default: none]
  -o OUTPUT   Output file [default: stdout]
  -r REPORT   Write classifier scores to file [default: none]
  -v          Verbose mode [default: False]
  -z          Apply gzip compression to output [default: False]
</pre></div>
</div>
</div>
<div class="section" id="creating-indexed-sorted-bam-files">
<h2>Creating indexed, sorted BAM files<a class="headerlink" href="#creating-indexed-sorted-bam-files" title="Permalink to this headline">¶</a></h2>
<p>If your reads are in the SAM format, you can convert it to a indexed,
sorted BAM file using <strong class="program">convertSam.py</strong>.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>usage: convertSam.py [-h] [-o BAMFILE] [-p PROCS] [-m MEMORY] [-v] samfile

Generate sorted BAM and index files for given SAM file

positional arguments:
  samfile               Samfile to convert

optional arguments:
  -h, --help            show this help message and exit
  -o BAMFILE, --outfile BAMFILE
                        Name of converted BAM file [default=&lt;sambase&gt;.bam]
  -p PROCS, --procs PROCS
                        Number of processors to use for BAM sorting (default
                        1)
  -m MEMORY, --memory MEMORY
                        Max memory (in GBs) for each processor used for BAM
                        sorting (default 2)
  -v, --verbose         Print verbose output
</pre></div>
</div>
<p>If your alignments are in a sorted BAM file, an index can be
generated using:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$samtools index &lt;BAMFILE&gt;
</pre></div>
</div>
<p>Both methods will produce .bai file, indicating that the BAM file
has been indexed.</p>
</div>
<div class="section" id="running-idiffir">
<h2>Running <strong>iDiffIR</strong><a class="headerlink" href="#running-idiffir" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ idiffir.py --help
usage: idiffir.py [-h] [-v] [-n] [-l FACTORLABEL FACTORLABEL] [-o OUTDIR] [-s]
                  [-k KRANGE KRANGE] [-c COVERAGE] [-d DEXPTHRESH]
                  [-f FDRLEVEL] [-g GRAPHDIRS] [-m {BF,BH,QV}] [-e {IR,SE}]
                  genemodel factor1Dirs factor2Dirs

Identify differentially expressed introns.

positional arguments:
  genemodel             gene model file: NAME.gtf[.gz] | NAME.gff[.gz]
  factor1Dirs           colon-separated list of directories: PATH-TO-
                        REPLICATE_1[:PATH-TO-REPLICATE_2,...]
  factor2Dirs           colon-separated list of directories: PATH-TO-
                        REPLICATE_1[:PATH-TO-REPLICATE_2,...]

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         verbose output [default is quiet running]
  -n, --noplot          Do not plot figures [default is to make figures]
  -l FACTORLABEL FACTORLABEL, --factorlabel FACTORLABEL FACTORLABEL
                        factor labels, example: -f Mutant Wildtype
  -o OUTDIR, --output-dir OUTDIR
                        output file directory name
  -s, --shrink_introns  shrink introns for depth plots [default is no
                        shrinking]
  -k KRANGE KRANGE, --krange KRANGE KRANGE
                        kmin kmax; [default is to search for kmax]
  -c COVERAGE, --coverage COVERAGE
                        coverage cutoff, default = 0.99
  -d DEXPTHRESH, --dexpThresh DEXPTHRESH
                        differential gene expression threshold, [default = 10]
  -f FDRLEVEL, --fdrlevel FDRLEVEL
                        FDR test level, [default = 0.05]
  -g GRAPHDIRS, --graph-dirs GRAPHDIRS
                        colon-separated list of directories to recursively
                        search for SpliceGrapher predictions
  -m {BF,BH,QV}, --multTest {BF,BH,QV}
                        Multiple testing adjustment method BF: Bonferroni, BH:
                        Benjamini-Hochberg, QV: q-values [default = QV]
  -e {IR,SE}, --event {IR,SE}
                        AS event to test, IR: Intron Retention, SE: Exon
                        Skipping [default = IR]
</pre></div>
</div>
<p>While <strong>iDiffIR</strong> offers many options, default values should work
for most cases.  Using the above naming convention, a way to run
<strong>iDiffIR</strong> would be:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>$idiffir.py -l Mutant Wildtype -o my_result my_genome.gff mt_rep1.bam:mt_rep2.bam:mt_rep3.bam wt_rep1.bam:wt_rep2.bam:wt_rep3.bam
</pre></div>
</div>
<p>The <strong>-l</strong> option provides titles for the reads and will be displayed
on the splice graphs of genes containing significant differential IR
events.    You may wish to use the <strong>-d</strong> option to lower the
gene expression fold-change threshold. Introns in genes
with a gene expression fold-change above this
threshold will not be tested.  While <strong>iDiffIR</strong> does
adjust for differential gene expression, in some extreme cases it may
be necessary.</p>
</div>
<div class="section" id="interpreting-idiffir-output">
<h2>Interpreting iDiffIR output<a class="headerlink" href="#interpreting-idiffir-output" title="Permalink to this headline">¶</a></h2>
<p><strong>iDiffIR</strong> builds an output directory as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ tree my_result
  my_result
  |-- figures
  |   |-- gene1.pdf
  |   |-- gene2.pdf
  |   |-- mva.pdf
  |   |-- pvalues.pdf
  |-- figuresLog
  |   |-- gene1.pdf
  |   |-- gene2.pdf
  |-- lists
      |-- allDIRGenes.txt
      |-- allIntrons.txt

  3 directories, 8 files
</pre></div>
</div>
<p>Two figures directories <strong>figures</strong> and <strong>figuresLog</strong> are created
and contain splice graphs figures of genes with at least
one significant
differential IR event.  The <strong>figuresLog</strong> directory contains
graphs in <span class="math">\(\log\)</span> scale which can be helpful in cases where
intronic expression is orders of magnitude less than the exonic
expression within the gene.  The <strong>lists</strong> directory contain
two important files.  <strong>allDIRGenes.txt</strong> contain the
gene IDs for all genes containing a significant differential IR
event.  This is should allow convenient downstream analyses,
such as GO term enrichment.  Finally, <strong>allIntrons.txt</strong> is a
tab-delimited file containing intronic coordinates and statistics
for all tested introns.  The fields of each row are:</p>
<ol class="arabic">
<li><dl class="first docutils">
<dt><strong>geneID</strong></dt>
<dd><p class="first last">the gene&#8217;s identifier</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>lowExonCoords</strong></dt>
<dd><p class="first last">the coordinates of the lowest (W.R.T. genomic
position) flanking exon</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>intronCoords</strong></dt>
<dd><p class="first last">the coordinates of the intronic region tested</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>highExonCoords</strong></dt>
<dd><p class="first last">the coordinates of the highest (W.R.T. genomic
position) flanking exon</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>pValue</strong></dt>
<dd><p class="first last">the <span class="math">\(p\)</span>-value of the tested intron (using a 2-sided
<span class="math">\(Z\)</span>-score test).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>adjPValue</strong></dt>
<dd><p class="first last">the multiple testing adjusted <cite>p</cite>-value</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>logFoldChange</strong></dt>
<dd><p class="first last">the <span class="math">\(\log\)</span>-fold change of the tested intron (w.r.t. the
first-given condition)</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>intronExp</strong></dt>
<dd><p class="first last">the expression of the intron, computed as
<span class="math">\(\displaystyle\frac{1}{2}\log\left( x_1 + x_2\right)\)</span>,
where <span class="math">\(x_1, x_2\)</span> are the average read depth of the intron
in condition 1 and condition 2, respectively.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>statistic</strong></dt>
<dd><p class="first last">the test statistic (before z-score conversion)</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>bestA</strong></dt>
<dd><p class="first last">the pseudo-count value (<span class="math">\(a\)</span>) that minimizes the
<span class="math">\(p\)</span>-value</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><strong>known</strong></dt>
<dd><p class="first last">whether this intron is known as a retained intron</p>
</dd>
</dl>
</li>
</ol>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">Add differential exon skipping</p>
</div>
<div class="admonition-todo admonition" id="index-1">
<p class="first admonition-title">Todo</p>
<p class="last">Add MISO script usage for testing Alt 5&#8217;, 3&#8217;</p>
</div>
<div class="admonition-todo admonition" id="index-2">
<p class="first admonition-title">Todo</p>
<p class="last">Add simulation</p>
</div>
<table class="docutils citation" frame="void" id="sg" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[SG]</a></td><td>Rogers, MF, Thomas, J, Reddy, AS, Ben-Hur, A (2012).
SpliceGrapher: detecting patterns of alternative splicing
from RNA-Seq data in the context of gene models and
EST data. <em>Genome Biol</em>., 13, 1:R4.</td></tr>
</tbody>
</table>
</div>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        «&#160;&#160;<a href="installation.html">Get iDiffIR</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="contact.html">Contact</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer">
        &copy; Copyright 2015, Michael Hamilton.
    </div>
  </body>
</html>